### ISSUE TITLE
framerate_producer: feature request

### ISSUE THREAD

**Issue by dimitry-ishenko** [CasparCG/Server#494]
I am looking at possibility of using the framerate_producer to implement jog/shuttle type features.

This is partially already possible with forward speeds between 0X and 1X. What's missing is the following:
1. Forward speeds above 1X. Eg, `call 1-10 framerate speed 8` (fast-forward at 8X)
2. Reverse speeds (for backwards playback). Eg, `call 1-10 framerate speed -0.25` (reverse slomo at 1/4X) or `call 1-10 framerate speed -8` (fast rewind at 8X)
3. Frame-stepping forward and backwards, while at speed 0.

@HellGore you think these are doable with the framerate_producer? If so, I can try digging into its source code to see if I can implement any of them. Will probably need some hand holding though.

**Comment by HellGore** [comment]
It require more modifications in `ffmpeg_producer` than in `framerate_producer`. `framerate_producer` should not be concerned with backwards playing, but `ffmpeg_producer` could provide such a feature.

If the `ffmpeg_producer` had a `CALL` telling it how many integer frames to advance (positive or negative) and to skip the decoding of frames in between you could instead of telling the `framerate_producer` go over speed 2.0, tell it to go down to 1, but tell the `ffmpeg_producer` to skip decoding of every other frame. Something like:

```
=========================================================================
framerate_producer speed  ffmpeg_producer advance_frame  Effective speed
------------------------- ------------------------------ ----------------
0.5                       +1                             0.5
1.0                       +1                             1.0
1.5                       +1                             1.5
1.0                       +2                             2.0
1.25                      +2                             2.5
1.0                       +3                             3.0
1.16667                   +3                             3.5
0.5                       -1                             -0.5
1.0                       -1                             -1.0
1.5                       -2                             -1.5
1.0                       -2                             -2.0
1.25                      -2                             -2.5
1.0                       -3                             -3.0
1.16667                   -3                             -3.5
=========================================================================

```

Then I guess you would also have to consider how to reverse files with long GOPs or if it will only work with i-frame only codecs.

**Comment by dimitry-ishenko** [comment]
@HellGore OK that makes sense. I can handle adding the call part. Where do I look to handle frame skipping?

**Comment by dimitry-ishenko** [comment]
It seems `caspar::ffmpeg::input::implementation::tick()` is the place where the frames get read from file. Is that the right place?

**Comment by dimitry-ishenko** [comment]
`ffmpeg_producer::try_decode_frame`?

**Comment by dimitry-ishenko** [comment]
@HellGore I was thinking there no real reason to do frame blending or fractional speeds above 2X. 
So, maybe we can do this:
- the `framerate_producer` will handle speeds below 2X
- the `ffmpeg_producer` will handle speeds starting from 2X and above and will ignore the fractional part.
This way they will not be "fighting" with each other to control the speed.

**Comment by dimitry-ishenko** [comment]
@HellGore so I played around with it a bit and came up with PR #497.

As explained in the message, I am discarding frames/samples to achieve higher speeds. This has some problems though:
- Audio sometimes goes out of sync with video. Seeking few seconds forward/backwards corrects it.
- I wasn't able to achieve speeds higher than 16X
- I've gotten audio buffer overflow once

Food for thought: Seeking might be a better alternative for this, considering it will be a must for reverse playback.

Any thoughts and feedback are welcome.

**Comment by dimitry-ishenko** [comment]
@HellGore you think it would be worthwhile trying the seek approach? I was thinking of basically adding speed function to `ffmpeg::input` class and when `speed_ != 1` calling `queued_seek` before `av_read_frame` from within `ffmpeg::input::tick`.
Will need to figure out correct value for `target` though...

**Comment by HellGore** [comment]
I have commented in the PR instead.

**Comment by dimitry-ishenko** [comment]
This is being addressed in #507
